infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java

```diff
@@ -301,29 +295,66 @@ public void acquireNode() {
+                                                                        vmSizeType,
+                                                                        resourceGroup,
+                                                                        region,
+                                                                        privateNetworkCIDR,
+                                                                        staticPublicIP);
+            LOGGER.info("Instances ids created or retrieved : " + instancesIds);
+        } else {
+            // if the infrastructure was already created, then wee need to
+            // look at the free instances, if any (the ones on which no node
+            // run. In the current implementation, this can only happen when
+            // nodes are down. Indeed if they are all removed on purpose, the
+            // instance should be shut down). Note that in this case, if the
+            // free instances map is empty, no script will be run at all.
+            Map<String, Integer> freeInstancesMap = getFreeInstancesMapCopy();
+            instancesIds = freeInstancesMap.keySet();
+            LOGGER.info("Instances ids previously saved which require script re-execution: " + instancesIds);
+            instanceIdSaved = true;
+        }
 
+        // execute script on instances to deploy or redeploy nodes on them
         for (String instanceId : instancesIds) {
-
             String fullScript = generateScriptFromInstanceId(instanceId);
-            connectorIaasController.executeScript(getInfrastructureId(), instanceId, Lists.newArrayList(fullScript));
+            try {
+                connectorIaasController.executeScript(getInfrastructureId(),
+                                                      instanceId,
+                                                      Lists.newArrayList(fullScript));
+            } catch (Exception e) {
+                // if we cannot execute the script although the infrastructure
+                // was already deployed, then it means that the Azure
+                // instances are probably dead, so we will attempt a
+                // redeployment from scratch
+                if (instanceIdSaved) {
+                    LOGGER.info("Saved instance: " + instanceId + " does not exist anymore. Recreating all instances.");
+                    clearFreeInstancesMap();
+                    compareAndSetAlreadyCreatedFlag(true, false);
+                    acquireNode();
+                    break;
+                } else {
+                    LOGGER.info("Script execution failed and cannot be handled, abandoning instance " + instanceId);
+                }
+            } finally {
+                // in all cases, we must remove the instance from the free
+                // instance map as we tried everything to deploy nodes on it
+                removeFreeInstanceFromMap(instanceId);
+            }
```

Comment:
All this catch/finally logic that redeploy the node couldn't be implemented the same way in OpenStack and VMWare?

Reply:
This is planed in longer term :) but first for EC2